### ISSUE TITLE
AE exporter 29.97 timebase differ from QT variant

### ISSUE THREAD

**Issue by exscriber** [GregBakker/adobe-cc-codec-foundation#12]
Timebase of 29.97 files from AE exporter plugin is 32768/982057,
and it differ from QT variant with expected 1001/30000

Diagnostic can be done with ffprobe:
`ffprobe -v error -show_format -show_streams 1.mov AE-PL-2997`
 `ffprobe -v error -show_entries stream=index,codec_time_base AE-PL-2997.mov`
Here is result of ffprobes: http://mergely.com/ccCLMh8A/

Issue not happens during AE-thru-AME export and not happens during normal QT exports.

So it seems Premiere and AfterEffects SDKs provide different timebase num/den values.
Looks like AE provide 1000/29970 (not 1001/30000) and then ffmpeg mess it even more...

Doesn`t find algorithm solution so far, only switch table with all timebase variants...

https://ae-plugin-sdk-guide.readthedocs.io/ppro/basic-host-differences.html#time-values

**Comment by GregBakker** [comment]
[also was late seeing this sorry! checking PR atm]

**Comment by GregBakker** [comment]
I'm not 100% understanding yet sorry!

1. What is the exact problem that's caused further downstream? Is the fps not displayed as 29.97 in disguise?

2. If AE is supplying a nonconformant num/den, it's still well-defined which frame should be displayed at which clock time - and replacing a num/den with a mathematically non-equal num/den may be problematic.

3. As per @bangnoise suggestion - is trashing the value from AE/Pr problematic, if the fraction is mathematically identical?

**Comment by exscriber** [comment]
There is no problems with files from AE. Disguise display both files as 29.97fps

Just some inconsistency when exporting directly from AE and thru AME.
btw it`s close to cosmetic, because difference is 1,000001

**Comment by GregBakker** [comment]
Ah I see; that's pretty annoying actually, they need to be consistant!

Maybe the problematic code is near here:
https://github.com/GregBakker/adobe-cc-codec-foundation/blob/366ed0dd726c80f6c5e8ffd3a0024185494e432d/source/plugin/premiere/premiereParams.cpp#L208

When an export is queued from AEX to AME, AEX probably has to dummy up a frame rate supported by the premiere exporter.

Maybe another solution would to be to add the AEX frame rates that we know about to the premiere exporter? This should allow them to passthrough unhindered. They *will* show up in the AME ui, but that's probably not a huge problem.

**Comment by exscriber** [comment]
Adobe's native QT (Hap or other codecs doesn`t matter):
timebase for 29/97 always 30000/1001: from AE direct export, AE thru AME, Premiere direct

CC-Hap plugins:
timebase 30000/1001: from Premiere direct and AE thru AME
timebase round(65536*2997/100)/65536: from AE direct export

I think Adobe does timebase conversion internally during AE thru AME, and it does same conversion during AE direct QT export.
So no changes needed to Premiere(AME) plugin, it uses Adobe's internal automagic... And it`s consistent with native QT.

With CC-Hap direct AE export there is no automagic, so it differ from others (very slight differ)

**Comment by GregBakker** [comment]
Ah ok. To summarise, we have two issues:
 1. exports specifically via the AE output module mangle the frame rate
 2. [not blocking but should be fixed] as Tom suggested, we should probably not divide by the gcd in MovieWriter and should just pass through. I believe we did that initially to deal with a reported fps issue; but later we found that a miscomputed duration caused the issue. We shouldn't really need to adjust a timebase in MovieWriter as we're not trying to pick a 'best' timebase to integrate multiple incoming timebases.

Just had a look, and the AE output module specifically has the problem as follows:

This contains the timebase:
https://github.com/GregBakker/adobe-cc-codec-foundation/blob/48e51fbe288eab1645812513b3175bc0cb6513c0/source/plugin/after_effects/output_module.cpp#L470

Obtained the timebase:
https://github.com/GregBakker/adobe-cc-codec-foundation/blob/48e51fbe288eab1645812513b3175bc0cb6513c0/source/plugin/after_effects/output_module.cpp#L494

Using the timebase:
https://github.com/GregBakker/adobe-cc-codec-foundation/blob/48e51fbe288eab1645812513b3175bc0cb6513c0/source/plugin/after_effects/output_module.cpp#L603-L604

The problem lies in the type returned by AEGP_GetOutSpecFPS. It's a fixed point number, with the defined denominator of A_Fixed_ONE, which in the AEX SDK header is #defined as ((A_Fixed)0x00010000L). This can't represent the native 30000/1001 directly.

The solution is to either find a better api to use to get the framerate [haven't had time to look for this yet!], or move your fix here from MovieWriter.
